### PR TITLE
Feature/sharedworker

### PR DESCRIPTION
**Pull request by adrienbarral**
**Public API Changes**
There is a new transportLibrary : sharedworker
There is a new transportOption : sharedWorkerURL

**Description**
A shared worker allow to create only One Websocket connection to rosbridge server for several browser windows (or tab).
This is like a worker (that run a script in a separate thread), but this worker is "shared" between all openned windows or tab.

You can try the exemple provided. If you open it in several windows of the same browser, there will be only one connection to the rosbridge server.

Of course, if you open one firefox and one chromium, you will have two connections.

I am not very good with Karma, so I did a little hack to make the test passing (I copy the worker script close to the test script). I tried to modify the karma.conf file to make it work without this copy, but I failed !

We've got a very big application that use roslibjs. During the next weeks, we will try this new feature in this application (I already did some tests on a smallest app). If we don't have any regressions, I will mark this PR as "mergeable". For now, I just publish it to have a first feed back.

Regards,

### CODE REVIEW

**Comment by MatthijsBurgh** [comment]
Before looking into the actual code, why are you coping the new file instead of including it in the build files?

**Comment by adrienbarral** [comment]
@MatthijsBurgh : If you speek about my hack for test. I precisely tried to do what you suggest. I add the ```build/sharedWorkerImpl.js" in the list of files that karma must serve for my tests. I tried to play with the "base" argument of the karma.conf.js file. But, unsuccessfully. I also look at this closed karma issue : https://github.com/karma-runner/karma/issues/1302... 

As I told in my comment, If the rest of the PR is good for you, I will work on this karma configuration.

**Comment by MatthijsBurgh** [comment]
I am talking about copying the source file to the build folder....

**Comment by adrienbarral** [comment]
Ok, 
I don't like to add this file too... But I think we need to do it.

At the begining, I tryied to use webworkify (like you did for the worker). You can look at  [this commit :](https://github.com/RobotWebTools/roslibjs/pull/524/commits/d539c8ff66fc6caba198d6f5e09203cab15ee5b7#)

But doing that, I learned that webworkify generate an unique URL each time we call it. So if in two browser tabs we call webworkify on the same resource, the URL will be differents. And so the browser can't understand that the resource we are fetching is the same file. So it instanciate a differents shared worker... So by using webworkify, everythings works, but our shared worker is not shared...

My conclusion is as follow : 
To make shared worker works we need to load them via the strictly same URL in all the tabs. This is why I added the new `transportOption.sharedWorkerURL`. In the exemple I provide in this PR, i set it to `../build/sharedWorkerSocketImpl.js`.
In my private application (based on angular cli), I add the sharedWorkerSocketImpl.js in the assets of my application.

**Comment by MatthijsBurgh** [comment]
For me the additional file that is required, is too big of a breaking point, to allow this feature.

**Comment by adrienbarral** [comment]
@MatthijsBurgh  : I succeed to make the shared worker works with the workerscript contained in the bundle.
The trick is here : https://github.com/RobotWebTools/roslibjs/pull/524/files/3d63be984c76c0291e09f6b4f58e77c971f75ea5..675a9066970fb45ee84086e8fad4337b2de3f8fe#diff-c8b8dc98173d711d9a5f56d84c366c58ca3bd253a3cc53ca5ef08ef8c2a5dda9R10

I use [brfs](https://www.npmjs.com/package/brfs). This is a small extension of browserify that allow to inline content of a file in a browserified javascript file. Then I create a data URL with the base64 encoded worker script file. Hence, URL is the same across all my browser context, and the shared worker is shared.

**Comment by MatthijsBurgh** [comment]
@adrienbarral Good job.

Could you please rebase on the current develop branch. So that GH actions can run.

File: package.json [hunk 1]
```diff
@@ -12,12 +12,13 @@
     "@xmldom/xmldom": "./src/util/shim/@xmldom/xmldom.js",
     "./src/util/decompressPng.js": "./src/util/shim/decompressPng.js"
   },
-  "devDependencies": {
+  "devDependencies": {
```
Comment (MatthijsBurgh) [inline package.json:15]: Fix whitespace

File: package.json [hunk 2]
```diff
@@ -12,12 +12,13 @@
...
     "./src/util/decompressPng.js": "./src/util/shim/decompressPng.js"
   },
-  "devDependencies": {
+  "devDependencies": {    
     "chai": "*",
     "grunt": "^1.0.0",
     "grunt-browserify": "^6.0.0",
     "grunt-cli": "^1.2.0",
     "grunt-contrib-clean": "^2.0.0",
+    "grunt-contrib-copy": "^1.0.0",
```
Comment (MatthijsBurgh) [inline package.json:21]: Not needed anymore, right?

File: src/util/sharedWorkerSocket.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            throw 'Unknown message from workersocket';
+        }
+    }
+};
+
+SharedWorkerConnection.prototype.close = function () {
+    this.worker_.port.postMessage({ type: 'CLOSE' });
+};
+
+module.exports = SharedWorkerConnection;
```
Comment (MatthijsBurgh) [inline src/util/sharedWorkerSocket.js:62]: Fix missing newline at EOF

File: src/util/sharedWorkerSocket.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    // Here we use brfs to inline the worker code in the script variable bellow.
+    // The following line is evaluated at build time (when browserify "compile" the file).
+    // See the doc : https://www.npmjs.com/package/brfs
+    var script = fs.readFileSync(__dirname + '/sharedWorkerSocketImpl.js', 'utf8');
+    var b64script = window.btoa(window.unescape(window.encodeURIComponent(script)));
+    /* To construct a shared worker we need an URL. This URL must be striclty the same in all
+       browser windows. Here URL is the base64 encoded version of the worker code, so for
+       a same worker code, the URL will be the same in all windows !
+     */
```
Comment (MatthijsBurgh) [inline src/util/sharedWorkerSocket.js:24]: Be consistent with comment style

File: test/sharedworkersocket/pubsub.example.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
+var expect = require('chai').expect;
+var ROSLIB = require('../..');
+
+describe('Topics Example', function() {
+    this.timeout(1000);
+
+
```
Comment (MatthijsBurgh) [inline test/sharedworkersocket/pubsub.example.js:7]: Too much whitespace

**Review by MatthijsBurgh — CHANGES_REQUESTED** [review]
First quick review, not really looked at the code.

**Comment by MatthijsBurgh** [comment]
You need to extend `test/build.bash` to include `test-sharedworkersocket`

File: test/sharedworkersocket/karma.conf.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
+module.exports = function(config) {
```
[the remark this answers is not in the document]
Comment (adrienbarral) [reply test/sharedworkersocket/karma.conf.js:1]: I agree, but this is a cut and paste of the already existing file : `test/workersocket/karma.conf.js`.
Do you want I clean boths ?

[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply test/sharedworkersocket/karma.conf.js:1]: Yes, we can clean whitespace without affecting the git history/blame

**Comment by MatthijsBurgh** [comment]
Please resolve the package-lock conflicts

**Comment by MatthijsBurgh** [comment]
@adrienbarral @lroixblue I have converted this to draft. Please fix the issues and just let me know when this PR is ready for a final review.

**Comment by lroixblue** [comment]
Thank you @MatthijsBurgh, we are still working on it. We will notify you when we will be ready for a final review.

**Comment by MatthijsBurgh** [comment]
@lroixblue please let me know, whether you have any plans to work on this. As we are getting close to making a final v1 release. After which more significant changes will be made. Which will make it impossible to merge this.

**Pull request closed without merging**